platforms/ide/ide-plugins/src/main/java/org/gradle/plugins/ide/internal/tooling/IsolatedProjectsSafeIdeaModelBuilder.java

```diff
@@ -0,0 +1,279 @@
+
+    private final IntermediateToolingModelProvider intermediateToolingModelProvider;
+    private final GradleProjectBuilderInternal gradleProjectBuilder;
+
+    public IsolatedProjectsSafeIdeaModelBuilder(IntermediateToolingModelProvider intermediateToolingModelProvider, GradleProjectBuilderInternal gradleProjectBuilder) {
+        this.intermediateToolingModelProvider = intermediateToolingModelProvider;
+        this.gradleProjectBuilder = gradleProjectBuilder;
+    }
+
+    @Override
+    public boolean canBuild(String modelName) {
+        return modelName.equals(MODEL_NAME);
+    }
+
+    @Override
+    public Class<IdeaModelParameter> getParameterType() {
+        return IdeaModelParameter.class;
+    }
+
+    @Override
+    public Object buildAll(String modelName, IdeaModelParameter parameter, Project project) {
+        return buildForRoot(project, parameter.getOfflineDependencyResolution());
+    }
+
+    @Override
+    public DefaultIdeaProject buildAll(String modelName, Project project) {
+        return buildForRoot(project, false);
+    }
+
+    @Override
+    public DefaultIdeaProject buildForRoot(Project project, boolean offlineDependencyResolution) {
+        Project root = project.getRootProject();
+        IdeaModelParameter parameter = createParameter(offlineDependencyResolution);
+        if (!root.equals(project)) {
+            return fetchForRoot(root, parameter);
+        }
+
+        // Ensure unique module names for dependencies substituted from included builds
+        applyIdeaProjectToBuildTree(root);
+
+        return buildRoot(root, parameter);
+    }
+
+    private DefaultIdeaProject fetchForRoot(Project root, IdeaModelParameter parameter) {
+        return intermediateToolingModelProvider.getModels(
+            Collections.singletonList(root),
+            MODEL_NAME,
+            DefaultIdeaProject.class,
+            parameter
+        ).get(0);
+    }
+
+    private void applyIdeaProjectToBuildTree(Project root) {
+        applyIdeaPluginToBuildTree((ProjectInternal) root, new ArrayList<>());
+    }
+
+    private void applyIdeaPluginToBuildTree(ProjectInternal root, List<GradleInternal> alreadyProcessed) {
+        intermediateToolingModelProvider.applyPlugin(new ArrayList<>(root.getAllprojects()), IdeaPlugin.class);
+
+        for (IncludedBuildInternal reference : root.getGradle().includedBuilds()) {
+            BuildState target = reference.getTarget();
+            if (target instanceof IncludedBuildState) {
+                GradleInternal build = target.getMutableModel();
+                if (!alreadyProcessed.contains(build)) {
+                    alreadyProcessed.add(build);
+                    applyIdeaPluginToBuildTree(build.getRootProject(), alreadyProcessed);
+                }
+            }
+        }
+    }
+
+    private DefaultIdeaProject buildRoot(Project rootProject, IdeaModelParameter parameter) {
+        rootProject.getPluginManager().apply(IdeaPlugin.class);
```

Comment:
🤔 Isn't this plugin application already covered by line 116?

Reply:
Originally, I thought we could find a solution that does not involve applying the plugin to every build tree project.
Alas, it's not that easy.

I don't feel strongly about this, but here application of a plugin is in the context of the builder for the current project, and it makes sense, because it picks up all extensions from the plugin. All the other applications does separately is a clutch that just has to be there and should eventually be removed. Since these two cases have separate reasons, I would keep the application here as well.

Reply:
I would at least add a comment here then:
// Even though this plugin application is already covered by ...